packages/core/src/lib/Cache.ts

```diff
@@ -4,158 +4,121 @@ import { ICategory } from "./Category.js"
 import { trace } from "@opentelemetry/api"
 import { Tags } from "../index.js"
 
-type Expiration = { absolute: number } | { relative: number }
-
-export class CacheEntry<State> {
+export class CacheEntry<T> {
   constructor(
     public token: StreamToken,
-    public state: State,
+    public state: T,
+    public cachedAt: number,
   ) {}
 
-  updateIfNewer(supersedes: (a: StreamToken, b: StreamToken) => boolean, other: CacheEntry<State>) {
+  updateIfNewer(other: CacheEntry<T>) {
     if (supersedes(this.token, other.token)) {
       this.token = other.token
       this.state = other.state
+      this.cachedAt = Date.now()
     }
   }
-  value(): TokenAndState<State> {
+  value(): TokenAndState<T> {
     return { token: this.token, state: this.state }
   }
 
   static ofTokenAndState<S>(x: TokenAndState<S>) {
-    return new CacheEntry(x.token, x.state)
+    return new CacheEntry(x.token, x.state, Date.now())
   }
 }
 
 export interface ICache {
-  updateIfNewer<State>(
+  readThrough<State>(
     key: string,
-    expiration: Expiration,
-    supersedes: (a: StreamToken, b: StreamToken) => boolean,
-    entry: CacheEntry<State>,
-  ): Promise<void>
+    skipReloadIfYoungerThanMs: number,
+    read: (tokenAndState?: TokenAndState<State>) => Promise<TokenAndState<State>>,
+  ): Promise<TokenAndState<State>>
 
-  tryGet<State>(key: string): Promise<TokenAndState<State> | undefined>
+  updateIfNewer<State>(key: string, entry: CacheEntry<State>): void
 }
 
-type Supersedes = (a: StreamToken, b: StreamToken) => boolean
+export const supersedes = (current: StreamToken, x: StreamToken) => x.version > current.version
```

Comment:
if an entry is re-read, you want to touch the entry if it is `=` (have not read full thing yet tho)